### ISSUE TITLE
Graph merge

### ISSUE THREAD

**Issue by m007** [neo4j-contrib/spatial#292]
## Use case
Suppose you already imported an .osm file into the graph. A use case for graph merge would be a scenario in which one would externally update this .osm file, and import this updated version to the graph database. The result of that operation should be an updated graph, with the updated data from the file (as opposed to previously deleting the old graph and creating a new one). 

## Issue
In #290 @craigtaverner wrote regarding graph merge:
> Back when we originally wrote the OSMImporter in 2011, we did start working on graph merge, but did not get that far with it, so I think it does not work at all at the moment. It is a non-trivial problem to solve, but is possible (with some work).

Since this is eventually unrelated to the original issue #290, this issue here is the place to discuss.

**Comment by m007** [comment]
I'd very much like to contribute to this, @craigtaverner. Would you provide some help to start?
At first. I'm interested in what the current state of the art is exactly and where to start...

**Comment by craigtaverner** [comment]
If I remember correctly, the thinking we had in 2011 was to separate the problem into two use cases:

1- merge multiple OSM files into single model where the OSM files represent multiple connected parts of the original OSM model (so are not different versions of same model). For example, import Sweden, then import Norway, and have them correctly connected. This requires identification of common elements (nodes, ways, relations) in the multiple OSM files. In practice it means if we load file1 first, then when loading file2 we correctly find existing nodes in the database for the incoming data and use them without creating new nodes or relationships. The current code was based on single once-off imports, so there would be changes in many places, but the current use of a lucene index to map from OSM node-id to Neo4j nodeId would be the foundation for the capability.

2- update an existing OSM model with new changesets from a newer version of the OSM model. This sounds more like the use case you described. We did work in this area by adding users and changesets to the Neo4j data model itself, so that we could identify if the incoming changes were already there and ignore them. However, the importer still focuses on adding nodes first and attaching changesets later, so we don't yet have a way to make that choice at the right place.

For this second case (which I think is the more complex case, but could be wrong), I think one option is to simply import the entire second model, and then do a model merge as a second step, removing all duplication. This would avoid the data ordering issue, but would be very inefficient, especially for the common case of the incoming model having only a few changesets different from the original.

To properly solve this issue, I think we should consider another important feature we planned to implement. And that is the complete re-write of the OSMImporter to use the new much faster parallel importer developed for Neo4j 2.1 or 2.2 back in 2014. That importer removed a very big bottleneck from the import process, which was the mapping of OSM nodeid to Neo4j nodeid. This slows down as the OSM file increases in size due to the use of a lucene index. The new generic importer solves this by doing a two phase import, firstly creating a table of neo4j nodes with pre-allocated neo4j node ids, and then importing those. This does require some more clever thinking for OSM, and possibly a multi-pass approach to parsing the OSM file. That will in turn allow us to do some more clever thinking about the nature of the OSM file and hopefully find and extract the new data (versus existing data) separately, and detect the best way to handle changesets.

**Comment by m007** [comment]
So yeah, what I had in mind and what you correctly guessed, is use case number two:
> 2- update an existing OSM model with new changesets from a newer version of the OSM model.

Summing up the state of the art for that:
To enable that feature, changesets & users were added to the graph, but since "the importer still focuses on adding nodes first and attaching changesets later" there is currently no way to use this information at the right time.

If I got that correctly, you do have two possible solutions for this in mind (and you're preferring 1. over 2.):
1. **A complete re-write of the OSMImporter** to use a "much more faster importer developed for Neo4j 2.1 or 2.2 back in 2014", which would "require some more clever thinking for OSM, and possibly a multi-pass approach to parsing the OSM file". Doing so, you think, will enable us to finally use the changeset & user information correctly to update an existing graph to a new OSM model.

2. Another approach to accomplish the above mentioned merging feature is to just **import the whole new OSM model and in a following step remove all duplicated nodes**. 

Did I get that correctly, or does 2. include a re-write of the OSMImpoter, too? If a re-write is needed, I may need some help, since that seems a rather big task.
